File: test/src/main/java/dev/akif/crud/InMemoryCRUDRepository.java
package dev.akif.crud;

import dev.akif.crud.error.CRUDErrorException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Map;
import java.util.Optional;
import java.util.function.BiPredicate;
import java.util.stream.Collectors;

/**
 * In-memory implementation of {@link CRUDRepository} that simulates an RDBMS for testing
 *
 * @param <I>      Id type of the data
 * @param <E>      Entity type of the data
 * @param <CM>     Create model type of the data
 * @param <Mapper> Mapper model type of the data
 */
public class InMemoryCRUDRepository<
        I extends Serializable,
        E extends CRUDEntity<I, E>,
        CM extends CRUDCreateModel<I, E>,
        Mapper extends CRUDTestMapper<I, E, ?, CM, ?>> implements CRUDRepository<I, E> {
    private final String type;
    private final Mapper mapper;
    private final BiPredicate<E, E> duplicateCheck;
    private final Map<I, E> entities;

    /**
     * Creates an in-memory repository containing given entities as initial data
     *
     * @param type            Type name of the data this repository manages
     * @param mapper          Mapper dependency of this repository
     * @param duplicate       Function to check whether two entities are duplicates
     * @param initialEntities Entities to set as initial data
     */
    @SafeVarargs
    public InMemoryCRUDRepository(final String type,
                                  final Mapper mapper,
                                  final BiPredicate<E, E> duplicate,
                                  final E... initialEntities
    ) {
        this.type = type;
        this.mapper = mapper;
        this.duplicateCheck = duplicate;
        this.entities = Arrays.stream(initialEntities).collect(Collectors.toMap(CRUDEntity::getId, e -> e));
    }

    @Override
    public Page<E> findAllByDeletedAtIsNull(final Pageable pageable) {
        return new PageImpl<>(
                entities
                        .values()
                        .stream()
                        .filter(e -> e.getDeletedAt() == null)
                        .skip(pageable.getOffset())
                        .limit(pageable.getPageSize())
                        .toList(),
                pageable,
                entities.size()
        );
    }

    @Override
    public Optional<E> findByIdAndDeletedAtIsNull(final I id) {
        return Optional.ofNullable(entities.get(id)).filter(e -> e.getDeletedAt() == null);
    }

    @Override
    public E save(final E entity) {
        handleDuplicates(entity);
        entities.put(entity.getId(), entity);
        return entity;
    }

    @Override
    public int updateByVersion(final E entity, final int version) {
        handleDuplicates(entity);
        return Optional.ofNullable(entities.get(entity.getId()))
                .filter(e -> e.getVersion() == version)
                .map(e -> entities.replace(entity.getId(), entity))
                .map(e -> 1)
                .orElse(0);
    }

    @Override
    public void flush() {
    }

    /**
     * Clears the repository for testing
     */
    public void clear() {
        entities.clear();
    }

    private void handleDuplicates(final E entity) {
        if (entities.values().stream().anyMatch(e -> duplicateCheck.test(entity, e))) {
            throw CRUDErrorException.alreadyExists(type, mapper.entityToCreateModel(entity));
        }
    }
}

File: test/src/main/kotlin/dev/akif/crud/InMemoryCRUDRepository.kt
package dev.akif.crud

import dev.akif.crud.error.CRUDErrorException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import java.io.Serializable

/**
 * In-memory implementation of [CRUDRepository] for testing
 *
 * @param I        Id type of the data
 * @param E        Entity type of the data which is a [CRUDEntity]
 * @param CM       Create model type of the data which is a [CRUDCreateModel]
 * @param TestData Test data type of the data which is a [CRUDTestData]
 *
 * @property testData Mapper dependency of this test which is a [CRUDTestData]
 */
open class InMemoryCRUDRepository<
        I : Serializable,
        E : CRUDEntity<I>,
        CM : CRUDCreateModel,
        TestData : CRUDTestData<I, E, *, CM, *, TestData>>(
    protected val testData: TestData
) : CRUDRepository<I, E> {
    private val log: Logger by lazy {
        LoggerFactory.getLogger(javaClass)
    }

    /**
     * Map of entities by their ids, accessible for tests
     */
    val entities: MutableMap<I, E> = mutableMapOf()

    init {
        reset()
    }

    override fun findAllByDeletedAtIsNull(pageable: Pageable): Page<E> =
        PageImpl(
            entities
                .values
                .stream()
                .filter { it.deletedAt == null }
                .skip(pageable.offset)
                .limit(pageable.pageSize.toLong())
                .map(testData::copy)
                .toList(),
            pageable,
            entities.count { (_, e) -> e.deletedAt == null }.toLong()
        )

    override fun findByIdAndDeletedAtIsNull(id: I): E? =
        entities[id]?.takeIf { it.deletedAt == null }?.let(testData::copy)

    override fun save(entity: E): E {
        handleDuplicates(entity, testData.entityToCreateModel(entity))
        if (entity.id == null) {
            entity.id = testData.idGenerator.next()
        }
        val copied = testData.copy(entity)
        entities[entity.id!!] = copied
        return copied
    }

    override fun update(entity: E): Int {
        handleDuplicates(entity, testData.entityToUpdateModelWithNoModifications(entity))
        val existing = entity.id?.let { entities[it] }?.takeIf { it.version == entity.version }
        val copied = testData.copy(entity).apply { version = version?.plus(1) }
        return existing
            ?.let { entity.id?.let { id -> entities.replace(id, copied) } }
            ?.let { 1 }
            ?: 0
    }

    override fun flush() {
        log.debug("Flushed repository: $entities")
    }

    /**
     * Clears the repository for testing, there will be no entities left after this.
     */
    fun clear() {
        entities.clear()
    }

    /**
     * Reset the repository for testing, there will be the initial entities after this.
     */
    fun reset() {
        entities.clear()
        entities[testData.testEntity1.id!!] = testData.testEntity1
        entities[testData.testEntity2.id!!] = testData.testEntity2
        entities[testData.testEntity3.id!!] = testData.testEntity3
        testData.moreTestEntities.forEach { testEntity ->
            entities[testEntity.id!!] = testEntity
        }
        log.debug("Reset repository: $entities")
    }

    private fun handleDuplicates(entity: E, data: Any) {
        val found = entities.values.any { e ->
            entity.deletedAt == null
                && e.deletedAt == null
                && testData.areDuplicates(entity, e)
        }
        if (found) {
            throw CRUDErrorException.alreadyExists(testData.typeName, data)
        }
    }
}
